### ISSUE TITLE
`_nll_loss_helper` is trying to index the output of `LogSoftmax` with an `ignore_index` value which is out of bounds!

### ISSUE THREAD

**Issue by kevinstephano** [Lightning-AI/lightning-thunder#1744]
## 🐛 Bug

_Note: I discovered this while trying to enable cross-entropy-loss through nvFuser.  The repro is purely through Thunder's torch executor._

In the `_nll_loss_helper` that decomposes `cross_entropy` loss, prior to index the output of `LogSoftmax` with the labels that are modified with an `ignore_index` of `-100`.  This index needs to be masked out because it is not in range of the vocabulary and results in an error as follows:

```bash
/opt/pytorch/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:144: operator(): block: [1,0,0], thread: [127,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
```

The issue is at the following lines:
https://github.com/Lightning-AI/lightning-thunder/blob/40f79727c419d3acf83cd75b51fa867810e23a46/thunder/torch/__init__.py#L5117-L5119

I think the `where` line needs to produce the target indices supplied to `take_along_dim`.

### To Reproduce

```bash
git clone https://github.com/kevinstephano/thunder_model_blocks.git
cd thunder_model_blocks
python hf_qwen2/qwen2_loss.py --execs Thunder-Torch
```

cc @tfogal

**Comment by kevinstephano** [comment]
The `-100` value is not specific to Qwen2.  It is the default value for adding a token as this failing for Phi3 and Mistral as well.

I think the code is shifting the labels to the right and adding an ignore index at the beginning of each sequence.
```
shifted_labels = [ignore_idx] + labels[1:]
```

**Comment by kevinstephano** [comment]
cc @protonu

**Comment by jjsjann123** [comment]
since we had all the issue with the repro. Turns out transformer change their loss function definition and I had an out-dated transformer installation.

Here's an easier repro.

```
import torch
 
batch_size = 2
seq_len = 4
vocab_size = 8
 
logits = torch.randn(batch_size * seq_len, vocab_size, device='cuda', dtype=torch.float32, requires_grad=True)
labels = torch.randint(0,vocab_size, (batch_size * seq_len, ), device='cuda', requires_grad=False)
 
# shift labels
labels = torch.nn.functional.pad(labels, (0,1), value=-100)[..., 1:].contiguous()
 
def fn(logits, labels):
    return torch.nn.functional.cross_entropy(logits, labels, ignore_index=-100)
 
ref_o = fn(logits, labels)
 
import thunder
from thunder.dynamo import thunderfx

thunder_fn = thunderfx(fn, executors=["torch"])
thunder_o = thunder_fn(logits, labels)
```

**Comment by jjsjann123** [comment]
We should discuss what is the proper approach. I think we need to update `prims.take_along_axis` to support a mode, like I initially suggested in the code comment.

But for the time being, #1752 would at least make it run without an error.
